Fix quadratic behavior of SparsityPattern::copy_from().

bangerth opened the pull request:
From the top of the test:
// Check SparsityPattern::copy_from(FullMatrix). This function took
// extraordinary amounts of CPU time: for a 1 x 400,000 matrix, it
// took around 50 seconds; after the patch that fixed the quadratic
// behavior, it takes less than one second
//
// To make sure we really catch the old bevahior through the test
// suite time out, we also do a matrix of size 1 x 4,000,000, which
// would have taken more than an hour before and now still takes less
// than one second. Finally, also try a square matrix because this
// case is treated specially in the function.

bangerth commented:
This is based on an observation by Ben Shields.

/run-tests

drwells commented on source/lac/sparsity_pattern.cc:
```diff
@@ -545,25 +545,59 @@ void SparsityPattern::copy_from (const FullMatrix<number> &matrix)
+  // gentler and hand over a whole row of entries at a time
+  std::vector<size_type> column_indices;
+  column_indices.reserve (*std::max_element (entries_per_row.begin(),
+                                             entries_per_row.end()));
```
somewhat pedantic, but this will segmentation fault for empty matrices, won't it?

bangerth replied:
Likely. Good spot -- this is why we do patch review!

I'll take care of this tomorrow and let you know once done!

kronbichler commented on source/lac/sparsity_pattern.cc:
```diff
@@ -545,25 +545,59 @@ void SparsityPattern::copy_from (const FullMatrix<number> &matrix)
   for (size_type row=0; row<matrix.m(); ++row)
-    for (size_type col=0; col<matrix.n(); ++col)
-      if (matrix(row,col) != 0)
-        add (row,col);
```
So the old quadratic complexity was because the `add` functions internally is of linear complexity? I've been trying to locate the quadratic complexity in the code here but did not find it.

drwells replied:
It is: `add` loops across the current row looking either for `j` or the next unused entry (which has value 2^64 - 1 or 2^32 - 1); to add the 4 millionth entry it reads the first 3,999,999 first :(

I tried changing `add` to add things in a sorted order with `lower_bound` and`rotate`, but it causes a massive number of tests to fail.

bangerth replied:
I had not actually looked into `add()` to understand why we got the quadratic behavior. I had just observed it experimentally. But @drwells 's explanation sounds correct.

bangerth commented:
I've fixed the segmentation fault when you work on empty matrices.

To make this work, I also had to relax a couple of assertions that check that certain functions can't be called on empty matrices. These seemed unreasonably restrictive, so I removed them. I then added another test for the copy operation with empty matrices.
